jerry-core/ecma/builtin-objects/ecma-builtin-helpers.c

```diff
@@ -588,27 +646,48 @@ ecma_builtin_helper_string_prototype_object_index_of (ecma_string_t *original_st
       {
         break;
       }
-      if (ecma_builtin_helper_string_find_index (original_str_p, search_str_p, true,
-                                                 (lit_utf8_size_t) start_ends_with, &index_of))
-      {
-        ret_value = ecma_make_boolean_value (index_of == (lit_utf8_size_t) start_ends_with);
-      }
+      uint32_t index = ecma_builtin_helper_string_find_index (original_str_p,
+                                                              search_str_p,
+                                                              (uint32_t) start_ends_with);
+      ret_value = ecma_make_boolean_value (index  == (uint32_t) start_ends_with);
       break;
     }
 #endif /* ENABLED (JERRY_ESNEXT) */
 
     case ECMA_STRING_INDEX_OF:
-    case ECMA_STRING_LAST_INDEX_OF:
-    default:
     {
       /* 8 (indexOf) -- 9 (lastIndexOf) */
-      if (ecma_builtin_helper_string_find_index (original_str_p, search_str_p, use_first_index, start, &index_of))
+      ecma_value_t find_index = ecma_builtin_helper_string_find_index (original_str_p, search_str_p, start);
+      if (find_index != UINT32_MAX)
       {
-        ret_num = ((ecma_number_t) index_of);
+        ret_num = ((ecma_number_t) find_index);
       }
       ret_value = ecma_make_number_value (ret_num);
       break;
     }
+
+    case ECMA_STRING_LAST_INDEX_OF:
+    {
+      uint32_t index = ecma_builtin_helper_string_find_last_index (original_str_p, search_str_p, start);
+
+      ecma_deref_ecma_string (search_str_p);
+
+      if (index == UINT32_MAX)
+      {
+        ret_value = ecma_make_number_value (ECMA_NUMBER_MINUS_ONE);
+      }
+      else
+      {
+        ret_value = ecma_make_uint32_value (index);
+      }
+
+      return ret_value;
```

Comment:
This should be a `break`, then the string deref on line 673 is no longer needed.